### PR TITLE
CSS url() rewriting middleware

### PR DESCRIPTION
**Pull request by evocateur**
This builds on @RCanine's work with the `rewriteCSSURLs` helper function by extracting it into an Express/Connect-style middleware function, while maintaining backward-compatibility when the `basePath` config is passed to the `combine` middleware.

I plan on letting this PR stew for at least a day, I have some comments to make about certain bits of code.

### CODE REVIEW

File: lib/combohandler.js [hunk 1]
```diff
@@ -31,6 +31,10 @@ exports.combine = function (config) {
         maxAge = 31536000; // one year in seconds
     }
 
+    if (basePath) {
+        rewriteCSSURLs = require('./middleware/cssUrls').rewrite;
```
Comment (evocateur) [inline lib/combohandler.js:35]: I envision this back-compat going away with a minor version bump (0.x.0) some time in the future. Or perhaps now, to avoid more clients implementing this pattern?

File: lib/combohandler.js [hunk 2]
```diff
@@ -57,6 +61,13 @@ exports.combine = function (config) {
             res.charset = 'utf-8';
             res.contentType(type);
 
+            // provide metadata to subsequent middleware via res.locals
+            res.locals({
+                relativePaths: query,
+                bodyContents: body,
```
Comment (evocateur) [inline lib/combohandler.js:67]: I'm not entirely happy with passing it along here and then a few lines later joining the array with newlines into `res.body`.

On the one hand, it's not impossible to detect if there is additional middleware, and even the identity of them, so I could optionally not join the array into `res.body`, relying on the subsequent middleware to do so. However, this is brittle, since we can't rely on everyone utilizing the new `combohandler#respond` endpoint middleware (which I would add an awareness of "is res.body an array?", etc).

So, anyway, I'm open to imrovements in this area.

File: lib/middleware/cssUrls.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
+/**
+Middleware that rewrites relative url() and import paths into
+absolute paths based on config (basePath).
+**/
+var path = require('path');
+var URI  = require("URIjs");
+
+var replacer = function replacer(basePath, relativePath) {
+    return function (substr, quote, match) {
+        var parsedURI = URI(match);
```
Comment (evocateur) [inline lib/middleware/cssUrls.js]: So it turns out URIjs is a pretty handy library. The [absoluteTo](http://medialize.github.io/URI.js/docs.html#absoluteto) method does exactly what we want, so I figured it would be a good idea to keep the code path as unified as possible.

File: lib/middleware/cssUrls.js [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        // this just does a simple sniff to validate whether the URL
+        // needs to be rewritten.
+        if (!parsedURI.is("relative")) {
+            return substr;
+        }
+
+        var fileBasePath = path.join(basePath, relativePath),
+            absolutePath = parsedURI.absoluteTo(fileBasePath);
+
+        return substr.replace(match, URI.decode(absolutePath));
```
Comment (evocateur) [inline lib/middleware/cssUrls.js]: Although, URIjs is _so_ strict that it was breaking the insanely escaped unit test (with all those psychotic backslash-escaped parens and quotes) because it URL-encodes everything like a good netizen. I'm torn on what is more correct: making the insane test case pass (is it based on real-world requirements, or just artifice?), or adhere closer to the RFCs and always URL-encode those kind of characters?

File: package.json [hunk 5]
```diff
@@ -25,7 +25,7 @@
 
     "dependencies": {
         "express": "3.1.1",
-        "URIjs"  : "1.10.0"
+        "URIjs"  : "1.10.1"
```
Comment (evocateur) [inline package.json:28]: There was a release last week that fixed an edge case with `absoluteTo`, so I figured it was worth bumping.

File: test/server.js [hunk 6]
```diff
@@ -321,5 +321,51 @@ describe('combohandler', function () {
                 done();
             });
         });
+
+        describe("as middleware", function () {
```
Comment (evocateur) [inline test/server.js:345]: I would seriously go insane without unit tests and code coverage.

**Comment by kara-ryli** [comment]
The reason why I moved the rewrites to array-based logic rather than using absoluteTo is that the substring you're matching isn't the URL, it's the _CSS-escaped URL_. Although the test is a bit contrived, an image with a non-ASCII character in the file name properly escaped is a real-world scenario, and using absoluteTo will make those files break.

I didn't want to capture all if the nuance of escapement in the CSS spec, so I went with the most minimal implementation I could with the least incompatibility with the spec. I think compatibility is more important here.

**Comment by evocateur** [comment]
So using the `path` utilities ends up being more robust, in my opinion, because it deals with all the edge cases of path construction (doubled slashes, no trailing slash, etc) for us without mangling any escapes.

However, [CSS character escapes](http://www.w3.org/International/questions/qa-escapes#cssescapes) are way tricksier than merely prefixing a punctuation character, and adding the following rules to urls.css exposes a shortcoming of the `url()` regex:

``` css
.unicode-raw { background: url(déchaîné.png); }
.unicode-escaped { background: url(d\E9 cha\EEn\E9.png); }
```

Namely, that space behind the first `\E9` that disambiguates it from the next character, `c`, which is otherwise valid hex (the first rule is fine, in a UTF8 document). Yes, you can work around it by providing the full six-digit hex (`\0000E9` in this case), and then the space is not needed. I'd rather avoid the complexity of parsing and recognizing and expanding hex escapes like this (or demanding authors use the "fully-qualified" variant), but it does mean that the regular expression needs to change.

I would like to replace the contrived escape example with the "real-world" examples I wrote above. We need to be able to match `([^\)]+)` in the url() regular expression, not `(\S+)`. I'll write some more test fixtures and push up the changes after a few meetings...

**Comment by evocateur** [comment]
I opted not to change the regex, and just added some test fixtures that include Unicode characters or their hex escapes, avoiding the space terminator. I'll be merging this PR by EOD today.

**Comment by rgrove** [comment]
:thumbsup:
